### PR TITLE
linkers: Add support for mold linker

### PR DESCRIPTION
**Pull request by Finii**
**[why]**
Support for the relatively new mold linker is missing. If someone wants
to use mold as linker `LDFLAGS="-B/path/to/mold"` has to be added instead
of the usual `CC_LD=mold meson ...` or `CXX_LD=mold meson ...`.

**[how]**
Allow `mold' as linker for clang and newer GCC versions (that versions
that have support).

The error message can be a bit off, because it is generic for all GNU
like compilers, but I guess that is ok. (i.e. 'mold' is not listed as
possible linker, even if it would be possible for the given compiler.)

**[note]**
GCC Version 12.0.1 is not sufficient to say `mold` is supported. The
expected release with support will be 12.1.0.
On the other hand people that use the un-released 12.0.1 will probably
have built it from trunk. Allowing 12.0.1 is helping bleeding edge
developers to use mold in Meson already now.

Fixes: #9072

_Edit: Adapt to force push_

### CODE REVIEW

**Comment by Finii** [comment]
Linking to the Issue with the commit message/PR description did not work ... `#` was missing :grimacing: 

Force push.

File: mesonbuild/compilers/mixins/gnu.py
Comment (dcbaker) [inline mesonbuild/compilers/mixins/gnu.py:405]: `cls` is the idiomatic name for referring to the `Type` object passed in a `classmethod`, while `self` is the idiomatic name of the `object` object passed to an instance method, so this comment isn't super useful.

Also, we shouldn't return something else here, we should error.

File: mesonbuild/linkers/linkers.py
Comment (dcbaker) [inline mesonbuild/linkers/linkers.py:793]: I'd rather not add mold here, the fact that mold acts so much like bfd and gold it can inherit that class directly is more of an implementation detail. I'm not entirely sure that it's safe to do that though, there may be cases where we use `isinstance(ld, GnuDynamicLinker)` that we don't want to find mold. It might makes more sense to copy the `LLVMDynaicLinker` hierarchy here, though we don't need to override anything it seems.

File: docs/markdown/Reference-tables.md
Comment (dcbaker) [inline docs/markdown/Reference-tables.md:49]: I know `"A Modern Linker"` is the slogan for mold, but most of these are more descriptive. Maybe a "High performance GNU like linker"?

**Review by dcbaker — CHANGES_REQUESTED** [review]
Thanks for working on this! I'd planned to get to it at some point but there's been so many other things on my plate. I have a few inline comments, but mostly this looks like the right direction.

This also needs a release snippet to be added to docs/markdown/snippets/

File: mesonbuild/compilers/mixins/gnu.py
Comment (Finii) [reply mesonbuild/compilers/mixins/gnu.py:402]: Removed/redesigned.

File: mesonbuild/compilers/mixins/gnu.py
Comment (Finii) [reply mesonbuild/compilers/mixins/gnu.py:405]: Of course you are right, I was kind of blind here.
Removed/redesigned.

File: mesonbuild/linkers/linkers.py
Comment (Finii) [reply mesonbuild/linkers/linkers.py:793]: Reverted.

File: docs/markdown/Reference-tables.md
Comment (Finii) [reply docs/markdown/Reference-tables.md:49]: Replaced by `The fast MOLD linker`.

I thought 'GNU like' is not saying anything specific (well it has the same interface, but unlike LLVM's linker it has just this one interface, so why mention it?

High performance was a bit too adverting speak for me, so downgraded to fast ;-)

I used the 'slogan' because that is the name/abbreviation expanded.

**Comment by Finii** [comment]
This needed a deeper redesign if we want to allow different linkers based on compiler version. Started afresh.

**Why**

The new linker `mold` is supported by some compiler, but only with recent versions of that compilers. That means knowledge about the compiler version is needed in `add_linker_args()`.

This is not possible with the current design in Meson: While preparing the instantiation of the compiler object we want to collect the linker arguments. But we do not have a finished compiler object in that moment, so we can not reference the version easily via `self.version` within the compiler object - we are in an un-initialized compiler class.

See approx https://github.com/mesonbuild/meson/blob/0.61/mesonbuild/compilers/detect.py#L525: While detecting the C compiler the function `guess_nix_compiler()` is called. Now that function gets/has no knowledge of the compiler version, but would need it to determine the linker arguments (that are allowed of that version of the compiler class `cls`).

On the other hand, the version information has already been determined at that point in time. It is just not possible to access it from within the compiler mixin that determines allowed linker arguments.

**How**

Hand down the version information as new/additional argument. The interface has been changed for all compilers and platforms to keep the interface unified (i.e. `guess_windows_compiler()`, ...) This is the reason why so many files are touched. In all instances we just hand down the exact same `version` string to the linker-finding-mechanism that is later used to construct the compiler object.

Afterwards the version number is trivially available in `add_linker_args()` and we can decide if `mold` is accepted or not.

**Comment by Finii** [comment]
Here a concise overview how the signature of existing functions were changed/expanded:
```diff
     @classmethod
-    def use_linker_args(cls, linker: str) -> T.List[str]:
+    def use_linker_args(cls, linker: str, version: str) -> T.List[str]:
```
```diff
 def guess_win_linker(env: 'Environment', compiler: T.List[str], comp_class: T.Type['Compiler'],
-                     for_machine: MachineChoice, *,
+                     comp_version: str, for_machine: MachineChoice, *,
                      use_linker_prefix: bool = True, invoked_directly: bool = True,
                      extra_args: T.Optional[T.List[str]] = None) -> 'DynamicLinker':
```
```diff
 def guess_nix_linker(env: 'Environment', compiler: T.List[str], comp_class: T.Type['Compiler'],
-                     for_machine: MachineChoice, *,
+                     comp_version: str, for_machine: MachineChoice, *,
                      extra_args: T.Optional[T.List[str]] = None) -> 'DynamicLinker':

```

**Review by dcbaker — APPROVED** [review]
LGTM, thanks for making those changes. I'll leave it up to @jpakkane whether this lands  for 0.62 or after the branchpoint

**Comment by Finii** [comment]
Further information:

It would be possible to enable `mold` also on older GCCs, we just need to set the correct path, like so:
```diff
@@ -399,6 +400,10 @@ class GnuCompiler(GnuLikeCompiler):
 
     @classmethod
     def use_linker_args(cls, linker: str, version: str) -> T.List[str]:
-        if linker == 'mold' and mesonlib.version_compare(version, '>=12.0.1'):
-            return ['-fuse-ld=mold']
+        if linker == 'mold':
+            if mesonlib.version_compare(version, '>=12.0.1'):
+                return ['-fuse-ld=mold']
+            linker_exe = shutil.which(linker)
+            if linker_exe:
+                return [f'-B/usr/local/libexec/mold']
         return super().use_linker_args(linker, version)
```
The problem is, that we need the path to the symlink to `mold` that is called `ld`. And that can be anything - at least I do not know any limits to search this or if there are already similar cases in Meson. Typically it will be either `/usr/libexec/mold` or `/usr/local/libexec/mold`, but I do not dare to hardcode these paths.

```
$ ll /usr/local/libexec/mold
total 0
lrwxrwxrwx 1 root root 19 Mär  8 18:18 ld -> /usr/local/bin/mold
```

And then I realized that Meson does not honor the `LDFLAGS` when it searches for the linker. That is some kind of **bug** I would say. If the user specifies a different linker via `LDFLAGS`, it is not used/found when Meson starts and prints the 'used' linker, but it is later when really building it is honored.

Of course it is possible to use `LDFLAGS` of the environment like we use `CC_LD` in the linker search process, but that seems to contradict the other changes that were implemented in the linker search related code. So ... I will not propose anything in this direction.

**Comment by Finii** [comment]
Another thing I found along the way, where I believe the code has an error...

image

```python
        if shutil.which(linker):
            if not shutil.which(linker):
                raise mesonlib.MesonException(
                    f'Cannot find linker {linker}.')
            return [f'-fuse-ld={linker}']
        return super().use_linker_args(linker, version)
```

Is it really intended to do these two `if`s? @dcbaker, at least for my understanding it can not be `true` and in the next line `None`. But it is in the same one commit.

**Comment by dcbaker** [comment]
I noticed that too looking at this. It's probably a botched rebase that was missed, I'd say it's safe to delete the `if not...` case altogether.

**Comment by dcbaker** [comment]
In the -B case, I'm happy to let users add `-Dc_link_args=-B...` for themselves and let the automatic detection solve it for them. There's simply too many variables to control for that to work reliably on our end.

**Comment by Finii** [comment]
How it looks "on my machine" with GCC build from `trunk` and also selfbuild `mold`:

image

**Comment by dcbaker** [comment]
That looks good, the failures on MSVC are expected, so I think once we get 0.62 out we'll get this merged

**Comment by xclaesse** [comment]
> That looks good, the failures on MSVC are expected, so I think once we get 0.62 out we'll get this merged

It has been released already, LGTM.

**Pull request merged**
